content/en/blog/_posts/2025-XX-XX-nftables-kube-proxy/index.md

```diff
@@ -0,0 +1,238 @@
+        # associated verdict.)
+	chain services {
+		ip daddr . meta l4proto . th dport vmap @service-ips
+	}
+
+        ...
+}
+```
+
+Since there's only a single rule, with a roughly **O(1)** map lookup,
+packet processing time is more or less constant regardless of cluster
+size, and the best/average/worst cases are very similar:
+
+![kube-proxy nftables first packet latency, at various percentiles, in clusters of various sizes (raw data available at link below)](nftables-only.svg)
+
+But note the huge difference in the vertical scale between the
+iptables and nftables graphs! In the clusters with 5000 and 10,000
+services, the average (p50) latency for nftables is about the same as
+the best-case (p01) latency for iptables. In the 30,000 service
+cluster, the _worst-case_ latency for nftables manages to beat out the
+_best-case_ latency for iptables by a few microseconds! We can combine
+the two graphs, but you may have to squint to see the nftables
+results!:
+
+![kube-proxy iptables-vs-nftables first packet latency, at various percentiles, in clusters of various sizes](iptables-vs-nftables.png)
+
+## Why NFTables? Part 2: Control Plane Latency
+
+While the improvements to data plane latency in large clusters are
+great, there's another problem with iptables kube-proxy that often
+keeps users from even being able to grow their clusters to that size:
+the time it takes kube-proxy to program new iptables rules when
+services and endpoints change.
+
+With both iptables and nftables, the total size of the "ruleset" as a
+whole (actual "rules", plus associated data) is **O(n)** in the combined
+number of services and endpoints. Originally, the iptables backend would
+rewrite every rule on every update, and with tens of thousands of
+services, this could grow to be hundreds of thousands of iptables rules.
+Starting in Kubernetes 1.26, we began improving kube-proxy so that it
+could skip updating _most_ of the unchanged rules in each update, but
+the limitations of `iptables-restore` as an API meant that it was still
+always necessary to send an update that's **O(n)** in the number of
+services (though with a noticeably smaller constant than it used to be).
+Even with those optimizations, it can still be necessary to make use of
+kube-proxy's `minSyncPeriod` config option to ensure that it doesn't
+spend every waking second trying to push iptables updates.
+
+The nftables APIs allow for doing much more incremental updates, and
+when kube-proxy in nftables mode does an update, the size of the
+update is only **O(n)** in the number of _changed_ services and
+endpoints. The fact that the nftables API allows each nftables-using
+component to have its own private table also means that there is no
+global lock contention between components like with iptables. As a
+result, kube-proxy's nftables updates can be done much more
+efficiently than with iptables.
+
+(Unfortunately I don't have cool graphs for this part.)
```

Comment:
there was one comparison I've done for programming time https://docs.google.com/presentation/d/1WMNZApX8HDHbi7ZnUFuy86UUYov22mBj/edit#slide=id.g30c6bb99704_1_0 
not much data, but maybe still useful?